### PR TITLE
fix(cli): Remove unsafe type assertions in activityLogger #19713

### LINKED ISSUE #19714: Fix unsafe type assertions in CLI Activity Logger (Phase 3.2)
Remove `no-unsafe-type-assertion` suppressions in `packages/cli/src/utils/activityLogger.ts`.

**Files:**
- `packages/cli/src/utils/activityLogger.ts`

**Pattern:**
- Casting `http` request/response objects and headers.

**Fix Strategy:**
- Use Node.js type definitions correctly (`http.IncomingMessage`, etc.).
- Avoid `any` in the patching logic if possible, or use type guards for headers.

### PR DESCRIPTION
**Pull request by Nixxx19**
## Summary

Removes all `no-unsafe-type-assertion` suppressions in `packages/cli/src/utils/activityLogger.ts` by replacing unsafe type assertions with strict type guards and runtime validation. This improves type safety when patching Node.js http/https request methods without changing functionality.

## Details

**Problem:**
The codebase had 9 `no-unsafe-type-assertion` suppressions where http request/response objects and headers were being cast without runtime validation, creating potential type safety issues.

**Solution:**
- **Type guards:** Added `isBufferEncoding()`, `isRequestOptions()`, and `isIncomingMessageCallback()` to validate types at runtime
- **Type-safe wrapper:** Created `callHttpRequest()` function that properly handles http.request overloads without unsafe assertions
- **Runtime checks:** All `as any`, `as http.RequestOptions`, and `as BufferEncoding` assertions replaced with proper runtime validation
- **Patching method:** Changed from `(http as any).request =` to `Object.defineProperty()` for safer module patching

**Pattern Fixed:**
- **Http request args:** Arguments to `http.request()` now validated before use using Node.js type definitions (`http.IncomingMessage`, `http.ClientRequest`, `http.RequestOptions`)
- **Buffer encoding:** Encoding parameters validated with type guard before use
- **Request options:** Options objects validated before passing to http.request
- **Http patching:** Replaced `as any` with `Object.defineProperty()` to avoid unsafe assertions in patching logic

**Technical Changes:**
- Added type guards: `isBufferEncoding()`, `isRequestOptions()`, `isIncomingMessageCallback()`
- Added wrapper function: `callHttpRequest()` for type-safe http.request calls
- Updated `wrapRequest()`: Uses type guards instead of assertions
- Updated http/https patching: Uses `Object.defineProperty()` instead of `as any`
- Removed all 9 `eslint-disable` comments for `no-unsafe-type-assertion`

## Related Issues

Fixes #19714

### Prerequisites
1. Build the project: `npm run build`

### Test Steps
1. **Verify no suppressions remain:**
   ```bash
   grep -r "eslint-disable.*no-unsafe-type-assertion" packages/cli/src/utils/activityLogger.ts
   ```
   **Expected:** No matches found

2. **Verify no unsafe assertions remain:**
   ```bash
   grep -r "as (any|http\.|BufferEncoding)" packages/cli/src/utils/activityLogger.ts
   ```
   **Expected:** No matches found (except one safe `as typeof http.request` for https compatibility)

3. **Run linter:**
   ```bash
   npm run lint -- packages/cli/src/utils/activityLogger.ts
   ```
   **Expected:** No linter errors related to unsafe type assertions
   
   **Note:** If the linter runs out of memory, you can verify by checking that there are no `no-unsafe-type-assertion` suppressions (see step 1).

4. **Run tests (if available):**
   ```bash
   npm test -- packages/cli/src/utils/activityLogger.ts
   ```
   **Expected:** All tests pass

5. **Verify Node.js types are used correctly:**
   ```bash
   grep -n "http\.\(IncomingMessage\|ClientRequest\|RequestOptions\)" packages/cli/src/utils/activityLogger.ts
   ```
   **Expected:** Multiple matches showing proper use of Node.js type definitions

### Expected Results
- All 9 `no-unsafe-type-assertion` suppressions removed
- All unsafe type assertions replaced with runtime validation
- Type guards properly validate types before use
- No linter errors
- Code maintains same functionality with improved type safety
- Node.js type definitions (`http.IncomingMessage`, `http.ClientRequest`, `http.RequestOptions`) used correctly throughout

### Edge Cases
- Test with invalid request args: Should be handled gracefully by `callHttpRequest()` with fallback to `{}`
- Test with missing properties: Should be handled by type guards
- Test with wrong types: Should be caught by type guards before use
- Test with invalid buffer encoding: Should return `undefined` instead of unsafe cast

## Pre-Merge Checklist

- [x] Updated relevant documentation and README (if needed)
- [ ] Added/updated tests (if needed) - No new tests needed, existing functionality preserved
- [x] Noted breaking changes (if any) - None
- [ ] Validated on required platforms/methods:
  - [x] MacOS
    - [x] npm run
    - [ ] npx
    - [ ] Docker
    - [ ] Podman
    - [ ] Seatbelt
  - [ ] Windows
    - [ ] npm run
    - [ ] npx
    - [ ] Docker
  - [ ] Linux
    - [ ] npm run
    - [ ] npx
    - [ ] Docker

### CODE REVIEW

File: packages/cli/src/utils/activityLogger.ts [hunk 1]
```diff
@@ -22,13 +22,152 @@ import WebSocket from 'ws';
...
+  }
+  if (args.length === 3) {
+    return (
+      (typeof first === 'string' || first instanceof URL) &&
+      typeof args[1] === 'object' &&
+      typeof args[2] === 'function'
+    );
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (Nixxx19) [reply packages/cli/src/utils/activityLogger.ts]: fixed that

File: packages/cli/src/utils/activityLogger.ts [hunk 2]
```diff
@@ -22,13 +22,92 @@ import WebSocket from 'ws';
...
 const MAX_BUFFER_SIZE = 100;
 
-/** Type guard: Array.isArray doesn't narrow readonly arrays in TS 5.8 */
 function isHeaderRecord(
   h: http.OutgoingHttpHeaders | readonly string[],
 ): h is http.OutgoingHttpHeaders {
   return !Array.isArray(h);
 }
 
+function isBufferEncoding(value: unknown): value is BufferEncoding {
```
Comment (gundermanc) [inline packages/cli/src/utils/activityLogger.ts]: I think a function already exists for this in the library. It's called `Buffer.isEncoding`

File: packages/cli/src/utils/activityLogger.ts [hunk 3]
```diff
@@ -545,12 +618,22 @@ export class ActivityLogger extends EventEmitter {
...
-    (https as any).request = (...args: unknown[]) =>
-      wrapRequest(originalHttpsRequest as typeof http.request, args, 'https:');
+    Object.defineProperty(http, 'request', {
+      value: (...args: unknown[]) =>
+        wrapRequest(originalRequest, args, 'http:'),
+      writable: true,
+      configurable: true,
+    });
+    Object.defineProperty(https, 'request', {
+      value: (...args: unknown[]) =>
```
Comment (gundermanc) [inline packages/cli/src/utils/activityLogger.ts]: It looks like this code is patching over the regular http.request() function? I think we can make this code a lot clearer by making the types here match what they are on the real function and cleaning up some of the unknowns.

**Review by gundermanc — CHANGES_REQUESTED** [review]

File: packages/cli/src/utils/activityLogger.ts (hunk 2 above)
Comment (Nixxx19) [reply packages/cli/src/utils/activityLogger.ts]: Thanks for the suggestion! I've updated the code to use `Buffer.isEncoding()` instead of the manual encoding checks. This is much cleaner and more maintainable.

File: packages/cli/src/utils/activityLogger.ts (hunk 3 above)
Comment (Nixxx19) [reply packages/cli/src/utils/activityLogger.ts]: Good catch! I've updated the type signatures to match the actual `http.request()` function signature, replacing `...args: unknown[]` with properly typed parameters (`url`, `options`, and `callback`). This provides much better type safety and clarity.

File: packages/cli/src/utils/activityLogger.ts (hunk 2 above)
Comment (Nixxx19) [reply packages/cli/src/utils/activityLogger.ts]: Done! Replaced the custom type guard with inline `Buffer.isEncoding()` checks.

File: packages/cli/src/utils/activityLogger.ts (hunk 3 above)
Comment (Nixxx19) [reply packages/cli/src/utils/activityLogger.ts]: Fixed! I've improved the type safety here:

- Replaced `unknown[]` with a proper `HttpRequestArgs` union type
- Typed chunk parameters as `string | Uint8Array` instead of `unknown`
- Updated the argument construction to only pass defined values (avoiding undefined in the arrays)

**Review by gundermanc — APPROVED** [review]
